### PR TITLE
Add gtest

### PR DESCRIPTION
**Pull request by cbachhuber**
Add gtest capability for unit testing.
- Gtest library is downloaded during cmake configuration. This is standard procedure and avoids needing to mess around with git submodules.
- Add unit testing to CI. The new job shares many setup steps with the existing ubuntu build job. I checked whether CI can reuse those steps between jobs, but that [does not seem to be possible](https://github.community/t5/GitHub-Actions/Support-for-YAML-anchors/td-p/30336). In future, we can probably outsource these setup steps to a common action to have at least a single source of truth (which is still executed multiple times in one CI cycle).
- I also tried to use docker images to speed up creation of the build environment, but I [didn't see a speed improvement](https://github.com/cbachhuber/dynamic-occupancy-grid-map/runs/599725115) compared to the current sequence of apt-get commands. Maybe if we use a slim docker image with only our dependencies, we can improve CI speed.
- Add a simple unit test for the Simulator to proof that everything above is working.

### CODE REVIEW

File: dogm/demo/utils/test/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ADD_EXECUTABLE (utils_spec simulator_spec.cpp)
```
Comment (TheCodez) [inline dogm/demo/utils/test/CMakeLists.txt:1]: To fix the windows build you should add `SET(gtest_force_shared_crt ON CACHE BOOL "" FORCE)` here.
See https://github.com/google/googletest/blob/master/googletest/README.md#visual-studio-dynamic-vs-static-runtimes

File: dogm/demo/utils/CMakeLists.txt [hunk 2]
```diff
@@ -28,4 +28,8 @@ TARGET_INCLUDE_DIRECTORIES(utils
     ${PROJECT_SOURCE_DIR}/include
     ${OpenCV_INCLUDE_DIRS})
 
-TARGET_LINK_LIBRARIES(utils PRIVATE ${OpenCV_LIBS})
\ No newline at end of file
+TARGET_LINK_LIBRARIES(utils PRIVATE ${OpenCV_LIBS})
+
+if(BUILD_TESTS)
```
Comment (TheCodez) [inline dogm/demo/utils/CMakeLists.txt:33]: I think it would be good for this to be an option. This way it can be set in cmake-gui.

**Comment by TheCodez** [comment]
Nice! 
Having problems integrating gtests was actually one of the reasons I never started the dogm test project.

I think the Ubuntu CI is fine for now. For Windows we might need to use Docker.

> Add unit testing to CI. The new job shares many setup steps with the existing ubuntu build job. I checked whether CI can reuse those steps between jobs, but that does not seem to be possible. In future, we can probably outsource these setup steps to a common action to have at least a single source of truth (which is still executed multiple times in one CI cycle).

Instead of two seperate jobs why not add the testing cmake step to the Ubuntu runner as a run step? Sure it would not be displayed as a seperate check on pull requests but clicking the Ubuntu check would reveal which checks failed anyways.

File: dogm/demo/utils/test/CMakeLists.txt (hunk 1 above)
Comment (cbachhuber) [reply dogm/demo/utils/test/CMakeLists.txt:3]: Done

File: dogm/demo/utils/CMakeLists.txt (hunk 2 above)
Comment (cbachhuber) [reply dogm/demo/utils/CMakeLists.txt:33]: Excellent idea! I set it to on per default.

**Comment by cbachhuber** [comment]
I also thought about just using one script for this. And I agree, let's just do it this way for now :+1:

**Comment by TheCodez** [comment]
Thank you!

**Pull request merged**